File: C++/rabinkarpstringmatch.cpp
#include <iostream>
using namespace std;

//Implements Rabin and Karp algorithm for efficient string matching

int* getOutput(string needle, string haystack);

void printOutput(int* output, int length)
{
	for (int i = 0; i < length; i++)
	{
		if (output[i] == 1) cout << "^";
		else cout << " ";
	}
}

int main()
{
  string needle;
  string haystack;

  getline(cin, needle);
  getline(cin, haystack);

  int length = haystack.length();
  int* output = getOutput(needle, haystack);

  cout << haystack << endl;
  printOutput(output, length);
  return 0;
}

int getHash(string needle)
{
	int val = 0;
	char c_val = 0;
	for (int i = 0; (unsigned)i < needle.length(); i++)
	{
		c_val = needle[i];
		val += c_val;
	}
	return val;
}

int* getOutput(string needle, string haystack)
{
	string output = "";
	int needleLength = needle.length();
	int haystackLength = haystack.length();

	int* positions = new int [haystackLength];

	for (int i = 0; i < haystackLength; i++)
	{
		positions[i] = 0;
	}

	int hashNeedle = getHash(needle);
	int hashHaystack = getHash(haystack.substr(0, needleLength));

	if (hashHaystack == hashNeedle)
	{
		if (haystack.substr(0, needleLength) == (needle))
		{
			positions[0] = 1;
		}
	}

	for (int i = 1; i <= (haystackLength - (needleLength)); i++)
	{
		hashHaystack -= (int)haystack[i-1];
		hashHaystack += (int)haystack[needleLength+i-1];
		if (hashHaystack == hashNeedle)
		{	
			if (haystack.substr(i, needleLength) == (needle))
			{
				positions[i] = 1;
			}
		}
	}
	
	return positions;
}

File: Python/rabinkarpstringmatch.py
#Implements Rabin and Karp algorithm for efficient string matching

def printOutput(output, length):
	for i in range(0, length):
		if output[i] == 1: 
			print('^', end="")
		else: 
			print(' ', end="")

def getHash(needle):
	val = 0
	c_val = 0
	for i in range(0, len(needle)):
		c_val = needle[i]
		val += ord(c_val)
	return val

def getOutput(needle, haystack):
	needleLength = len(needle)
	haystackLength = len(haystack)

	positions = [0]*haystackLength

	for i in range(0, haystackLength):
		positions[i] = 0

	hashNeedle = getHash(needle)
	hashHaystack = getHash(haystack[0:needleLength])

	if (hashHaystack == hashNeedle):
		if (haystack[:needleLength] == (needle)):
			positions[0] = 1

	for i in range (0, (haystackLength - needleLength)+1):
		c_val = haystack[i-1]
		val = ord(c_val);
		hashHaystack -= val
		c_val = haystack[needleLength+i-1]
		val = ord(c_val);
		hashHaystack += val
		if (hashHaystack == hashNeedle):
			if (haystack[i:needleLength+i] == (needle)):
				positions[i] = 1
	
	return positions

needle = input()
haystack = input()

length = len(haystack)

output = getOutput(needle, haystack)
print(haystack)
printOutput(output, length)